csrc/scheduler/greedy.cpp

```diff
@@ -307,32 +340,41 @@ class CompileTimeChecker : private IterVisitor {
     // For the scatter in and out tensors, the scatter dimension must
     // not be parallelized with TID.
     auto out_tv = ir_utils::getTvOutput(scatter);
-    checkConstrainedTv(out_tv, {constrained_out_logical_dim});
+    checkDomainConstraints(
+        out_tv->getLoopDomain(), {constrained_out_logical_dim});
 
     // In addition, the index and src tensors are not allowed to use
     // TID with the scatter dim. Their logical domains are not mapped
     // with the logical domains of the input and output tensors, so
     // they need to be checked separately.
-    checkConstrainedTv(
-        scatter->index()->as<TensorView>(), {constrained_out_logical_dim});
+    checkDomainConstraints(
+        scatter->index()->as<TensorView>()->getLogicalDomain(),
+        {constrained_out_logical_dim});
     // Index and src tensors are mapped, so just checking index should
     // be sufficient.
   }
 
   void handle(PadOp* pad) override {
-    checkConstrainedTv(ir_utils::getTvOutput(pad), pad->getPaddedAxes());
+    checkDomainConstraints(
+        ir_utils::getTvOutput(pad)->getLogicalDomain(), pad->getPaddedAxes());
   }
 
   void handle(TopKOp* topk) override {
     // Due to the current limitations of TopKOp codegen, all of the
     // TIDx threads participate, which means the TID parallelized iter
     // domain of this TopKOp must have an extent that is no less
-    // than any other TID parallelized iter domains. Requiring the
-    // exactness is not necessary but sufficient.
-    needs_exact_block_dim_ = true;
+    // than any other TID parallelized iter domains.
+    needs_all_tid_participation_ = true;
 
+    auto in_tv = ir_utils::getTvInput(topk);
     auto out_tv = ir_utils::getTvOutput(topk);
-    checkConstrainedTv(out_tv, {topk->dim()});
+
+    // Like ScatterOp, the input defines the scheduling, so check the
+    // input logical domain
+    checkDomainConstraints(
```

Comment:
TopKOp is in some sense similar to ScatterOp since the output logical shape actually doesn't define the loop extent but the input extent must be small enough to fit.